compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -2043,9 +2043,9 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
 
                         let ty_fn_ptr_from = tcx.safe_to_unsafe_fn_ty(fn_sig);
 
-                        if let Err(terr) = self.eq_types(
-                            *ty,
+                        if let Err(terr) = self.sub_types(
                             ty_fn_ptr_from,
+                            *ty,
```

Comment:
can you flip the args and check whether there's an existing test which goes from fail to pass to make sure we check for that potential unsoundness?

r=me after that

Reply:
Yes, `tests/ui/nll/mir_check_cast_unsafe_fn.rs` checks this, which was added 7 years ago:

```
#![allow(dead_code)]

fn bar<'a>(input: &'a u32, f: fn(&'a u32) -> &'a u32) -> &'static u32 {
    // Here the NLL checker must relate the types in `f` to the types
    // in `g`. These are related via the `UnsafeFnPointer` cast.
    let g: unsafe fn(_) -> _ = f;
    unsafe { g(input) }
    //~^ ERROR lifetime may not live long enough
}

fn main() {}
```